src/eth/executor.rs

```diff
@@ -72,42 +71,77 @@ impl EthExecutor {
     }
 
     /// Imports an external block using the offline flow.
-    pub async fn import_offline(&self, block: ExternalBlock, receipts: &[&ExternalReceipt]) -> anyhow::Result<()> {
-        tracing::info!(number = %block.number(), "importing offline block");
+    pub async fn import_offline(&self, external_block: ExternalBlock, external_receipts: &[&ExternalReceipt]) -> anyhow::Result<()> {
+        tracing::info!(number = %external_block.number(), "importing offline block");
 
         // index receipts
-        let mut receipts_by_hash = HashMap::with_capacity(receipts.len());
-        for receipt in receipts {
+        let mut receipts_by_hash: HashMap<H256, &ExternalReceipt> = HashMap::with_capacity(external_receipts.len());
+        for receipt in external_receipts {
             receipts_by_hash.insert(receipt.transaction_hash, receipt);
         }
 
+        let mut executions = Vec::new();
+
         // re-execute transactions
-        for transaction in block.transactions.clone() {
-            self.execute_in_evm(transaction.into()).await?;
+        for transaction in external_block.transactions.clone() {
+            let tx_receipt = receipts_by_hash
+                .get(&transaction.hash)
+                .ok_or(anyhow!("receipt not found for transaction {}", transaction.hash))?;
+
+            let input = transaction.clone().into();
+            let execution = self.execute_in_evm(input).await?;
+
+            execution.cmp_with_receipt(tx_receipt);
+
+            let transaction_input = transaction.try_into().or(Err(anyhow!("failed to convert tx input")))?;
+            executions.push((transaction_input, execution));
         }
 
+        let block = if let Some(executions) = NonEmpty::from_vec(executions) {
+            self.miner.lock().await.mine_with_many_transactions(executions).await?
```

Comment:
There is no need for miner here.
Miner will not be used because we are not going to generate new blocks in this flow.
We will import external blocks as-is, keeping the same numbers and transactions.